### ISSUE TITLE
Support exec

### ISSUE THREAD

**Issue by samuelkarp** [firecracker-microvm/firecracker-containerd#77]
We need to be able to support `task.Exec` for executing additional processes inside the containers.

**Comment by samuelkarp** [comment]
One challenge here will be managing port assignment for stdio vsocks, or fifo assignment for in-VM processing.

**Comment by mxpv** [comment]
> One challenge here will be managing port assignment for stdio vsocks

We might want to consider another approach for this, using grpc. This eliminates headache with ports.
```
service Agent {
	rpc WriteStdin(WriteStreamRequest) returns (WriteStreamResponse);
	rpc ReadStdout(ReadStreamRequest) returns (ReadStreamResponse);
	rpc ReadStderr(ReadStreamRequest) returns (ReadStreamResponse);
}
```

**Comment by IRCody** [comment]
>We might want to consider another approach for this, using grpc. This eliminates headache with ports.

>>service Agent {
	rpc WriteStdin(WriteStreamRequest) returns (WriteStreamResponse);
	rpc ReadStdout(ReadStreamRequest) returns (ReadStreamResponse);
	rpc ReadStderr(ReadStreamRequest) returns (ReadStreamResponse);
}

Is the idea here to multi-plex the streams into a single one?

**Comment by mxpv** [comment]
>  gRPC marshalling/unmarshalling 
We send raw bytes buffer, I don't think marshaling will be an issue here.

> Is the idea here to multi-plex the streams into a single one?
Not really. I think we can specify `container id` in `ReadStreamRequest`, so you still have 3 streams, but instead of making 3 vsock connections you just call ReadStdout/ReadStderr/WriteStdin

**Comment by IRCody** [comment]
Something like that could work but will container id be specific enough in the case of exec where you'd end up with multiple streams from the same container?

ttrpc doesn't support streams AFAIK so we'd have to mix ttrpc and gRPC which I can see being a little confusing.

**Comment by mxpv** [comment]
> Something like that could work but will container id be specific enough in the case of exec where you'd end up with multiple streams from the same container?

we can have `container id` + `exec id`

**Comment by samuelkarp** [comment]
> We send raw bytes buffer, I don't think marshaling will be an issue here.

@mxpv Can you explain this a bit more?  I'm not following; gRPC is based on protobuf and so stdout/stderr would need to be marshalled into a protobuf structure before being written over the vsock.  And if nothing else, there's the overhead of HTTP2, which gRPC requires but we could elide if we directly put bytes on the vsock.

**Comment by mxpv** [comment]
> gRPC is based on protobuf and so stdout/stderr would need to be marshalled into a protobuf structure before being written over the vsock.

There will be an overhead when packing protobuf data, but for bytes buffer it'll be something like:
```go
func Marshal() outBuf {
  if len(buf) > 0 {
    outBuf += bufSize
    outBuf += buf
}
}
```

Here is more real example: [1](https://github.com/kata-containers/agent/blob/a581aebf47386df570f796cb835bd23820448b9b/protocols/grpc/agent.pb.go#L977), [2](https://github.com/kata-containers/agent/blob/a581aebf47386df570f796cb835bd23820448b9b/protocols/grpc/agent.pb.go#L3878)

I don't think this overhead will be an issue.

To summarize: 

Here are problems with vsock impl that I see:
- Need to manage vsock ports
- Need somehow exchange this info with Agent
- Dealing with low-level sockets/connections/retries inconvenient.

Problem with GRPC approach:
- Protobuf marshaling overhead

**Comment by IRCody** [comment]
> * Need to manage vsock ports
> * Need somehow exchange this info with Agent

We are talking about gRPC over vsock right? Seems like you'd have to do these either way.

**Comment by mxpv** [comment]
We need just one well known vsock port and GRPC will handle connections.
